server/src/main/java/org/opensearch/cluster/metadata/Metadata.java

```diff
@@ -1425,6 +1429,55 @@ public Builder generateClusterUuidIfNeeded() {
         }
 
         public Metadata build() {
+            TimeValue buildStartTime = TimeValue.timeValueMillis(System.nanoTime());
+            DataStreamMetadata dataStreamMetadata = (DataStreamMetadata) this.customs.get(DataStreamMetadata.TYPE);
+            DataStreamMetadata previousDataStreamMetadata = (previousMetadata != null)
+                ? (DataStreamMetadata) this.previousMetadata.customs.get(DataStreamMetadata.TYPE)
+                : null;
+
+            boolean recomputeRequired = (previousMetadata == null)
+                || (indices.equals(previousMetadata.indices) == false)
+                || (previousDataStreamMetadata != null && previousDataStreamMetadata.equals(dataStreamMetadata) == false)
+                || (dataStreamMetadata != null && dataStreamMetadata.equals(previousDataStreamMetadata) == false);
+            TimeValue recomputeEndTime = TimeValue.timeValueMillis(System.nanoTime());
+            logger.info(
+                "Recompute required: {}, time taken for comparing indices: {} ms",
+                recomputeRequired,
+                (recomputeEndTime.getNanos() - buildStartTime.getNanos()) / 1000000L
+            );
+
+            Metadata metadata = (recomputeRequired == false)
+                ? buildMetadataWithPreviousIndicesLookups()
+                : buildMetadataWithRecomputedIndicesLookups();
+            TimeValue endBuildTime = TimeValue.timeValueMillis(System.nanoTime());
+            // Logging for testing only - will remove in future iterations
+            logger.info("built metadata in {} ms", (endBuildTime.millis() - buildStartTime.millis()) / 1000000L);
+            return metadata;
+        }
+
+        private Metadata buildMetadataWithPreviousIndicesLookups() {
+            return new Metadata(
+                clusterUUID,
+                clusterUUIDCommitted,
+                version,
+                coordinationMetadata,
+                transientSettings,
+                persistentSettings,
+                hashesOfConsistentSettings,
+                indices,
+                templates,
+                customs,
+                Arrays.copyOf(previousMetadata.allIndices, previousMetadata.allIndices.length),
+                Arrays.copyOf(previousMetadata.visibleIndices, previousMetadata.visibleIndices.length),
+                Arrays.copyOf(previousMetadata.allOpenIndices, previousMetadata.allOpenIndices.length),
+                Arrays.copyOf(previousMetadata.visibleOpenIndices, previousMetadata.visibleOpenIndices.length),
+                Arrays.copyOf(previousMetadata.allClosedIndices, previousMetadata.allClosedIndices.length),
+                Arrays.copyOf(previousMetadata.visibleClosedIndices, previousMetadata.visibleClosedIndices.length),
+                new TreeMap<>(previousMetadata.indicesLookup)
```

Comment:
You need to apply `Collections.unmodifiableSortedMap` to make the map unmodifiable.

Reply:
Thanks for pointing, I will revise it in my next iteration, once I get reviews on this approach.